File: java/com/google/dotprompt/models/ParsedPrompt.java
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.google.dotprompt.models;

import java.util.List;
import java.util.Map;

/**
 * Represents a prompt after parsing its metadata and template.
 *
 * <p>This record extends the concept of {@link PromptMetadata} by adding the template string with
 * frontmatter removed. It corresponds to the TypeScript {@code ParsedPrompt} interface.
 *
 * <p>Example:
 *
 * <pre>{@code
 * String source = """
 *     ---
 *     model: gemini-1.5-pro
 *     config:
 *       temperature: 0.7
 *     ---
 *     Hello {{name}}!
 *     """;
 * ParsedPrompt parsed = Parser.parseDocument(source);
 * // parsed.template() -> "Hello {{name}}!"
 * // parsed.model() -> "gemini-1.5-pro"
 * }</pre>
 *
 * @param template The source of the template with metadata/frontmatter already removed.
 * @param name The name of the prompt.
 * @param variant The variant name for the prompt.
 * @param version The version of the prompt.
 * @param description A description of the prompt.
 * @param model The name of the model to use.
 * @param tools Names of tools to allow use of in this prompt.
 * @param toolDefs Definitions of tools to allow use of in this prompt.
 * @param config Model configuration.
 * @param input Configuration for input variables.
 * @param output Configuration for output format.
 * @param raw The raw frontmatter as parsed.
 * @param ext Extension fields gathered by namespace.
 * @param metadata Arbitrary metadata.
 */
public record ParsedPrompt(
    String template,
    String name,
    String variant,
    String version,
    String description,
    String model,
    List<String> tools,
    List<ToolDefinition> toolDefs,
    Map<String, Object> config,
    PromptMetadata.InputConfig input,
    PromptMetadata.OutputConfig output,
    Map<String, Object> raw,
    Map<String, Map<String, Object>> ext,
    Map<String, Object> metadata) {

  /**
   * Creates a ParsedPrompt with just a template and no metadata.
   *
   * @param template The template string.
   */
  public ParsedPrompt(String template) {
    this(template, null, null, null, null, null, null, null, null, null, null, null, null, null);
  }

  /**
   * Creates a ParsedPrompt from a template and PromptMetadata.
   *
   * @param template The template string (frontmatter removed).
   * @param metadata The parsed metadata.
   * @return A new ParsedPrompt instance.
   */
  public static ParsedPrompt fromMetadata(String template, PromptMetadata metadata) {
    if (metadata == null) {
      return new ParsedPrompt(template);
    }
    return new ParsedPrompt(
        template,
        metadata.name(),
        metadata.variant(),
        metadata.version(),
        metadata.description(),
        metadata.model(),
        metadata.tools(),
        metadata.toolDefs(),
        metadata.config(),
        metadata.input(),
        metadata.output(),
        metadata.raw(),
        metadata.ext(),
        metadata.metadata());
  }

  /**
   * Converts this ParsedPrompt to a PromptMetadata (without the template).
   *
   * @return A PromptMetadata instance with the same metadata fields.
   */
  public PromptMetadata toMetadata() {
    return new PromptMetadata(
        name,
        variant,
        version,
        description,
        model,
        tools,
        toolDefs,
        config,
        input,
        output,
        raw,
        ext,
        metadata);
  }
}

File: dart/dotprompt/lib/src/models/parsed_prompt.dart
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

/// Parsed prompt representation.
///
/// Contains both the raw template string and parsed metadata from the
/// YAML frontmatter.
library;

import "package:meta/meta.dart";

import "prompt_metadata.dart";

/// A parsed prompt with template and metadata.
///
/// This represents a prompt file after parsing the YAML frontmatter,
/// containing both the raw template body and the structured metadata.
///
/// ## Example
///
/// ```dart
/// final source = '''
/// ---
/// model: gemini-pro
/// config:
///   temperature: 0.7
/// ---
/// Hello {{name}}!
/// ''';
///
/// final parsed = Parser.parseDocument(source);
/// print(parsed.template);  // "Hello {{name}}!"
/// print(parsed.model);     // "gemini-pro"
/// ```
@immutable
class ParsedPrompt {
  /// Creates a new [ParsedPrompt].
  const ParsedPrompt({required this.template, required this.metadata});

  /// Creates a [ParsedPrompt] from the template and metadata.
  factory ParsedPrompt.fromMetadata(String template, PromptMetadata metadata) =>
      ParsedPrompt(template: template, metadata: metadata);

  /// Creates a [ParsedPrompt] from a JSON map.
  factory ParsedPrompt.fromJson(Map<String, dynamic> json) => ParsedPrompt(
        template: json["template"] as String,
        metadata: PromptMetadata.fromJson(
          json["metadata"] as Map<String, dynamic>? ?? {},
        ),
      );

  /// The template body (after frontmatter).
  final String template;

  /// The parsed metadata.
  final PromptMetadata metadata;

  /// The model name from metadata.
  String? get model => metadata.model;

  /// The generation config from metadata.
  Map<String, dynamic>? get config => metadata.config;

  /// The tools list from metadata.
  List<String>? get tools => metadata.tools;

  /// The input schema from metadata.
  InputConfig? get input => metadata.input;

  /// The output config from metadata.
  OutputConfig? get output => metadata.output;

  /// The resolved tool definitions.
  List<ToolDefinition>? get toolDefs => metadata.toolDefs;

  /// Extension fields from metadata.
  Map<String, Map<String, dynamic>>? get ext => metadata.ext;

  /// Raw frontmatter as parsed.
  Map<String, dynamic>? get raw => metadata.raw;

  /// Converts this to a JSON-serializable map.
  Map<String, dynamic> toJson() => {
        "template": template,
        "metadata": metadata.toJson(),
      };

  @override
  bool operator ==(Object other) =>
      identical(this, other) || other is ParsedPrompt && template == other.template && metadata == other.metadata;

  @override
  int get hashCode => Object.hash(template, metadata);

  @override
  String toString() => "ParsedPrompt(template: $template, metadata: $metadata)";
}

/// A tool definition.
@immutable
class ToolDefinition {
  /// Creates a new [ToolDefinition].
  const ToolDefinition({
    required this.name,
    this.description,
    this.inputSchema,
    this.outputSchema,
  });

  /// Creates a [ToolDefinition] from a JSON map.
  factory ToolDefinition.fromJson(Map<String, dynamic> json) => ToolDefinition(
        name: json["name"] as String,
        description: json["description"] as String?,
        inputSchema: json["inputSchema"] as Map<String, dynamic>?,
        outputSchema: json["outputSchema"] as Map<String, dynamic>?,
      );

  /// The tool name.
  final String name;

  /// A description of what the tool does.
  final String? description;

  /// JSON Schema for the tool's input.
  final Map<String, dynamic>? inputSchema;

  /// JSON Schema for the tool's output.
  final Map<String, dynamic>? outputSchema;

  /// Converts this to a JSON-serializable map.
  Map<String, dynamic> toJson() => {
        "name": name,
        if (description != null) "description": description,
        if (inputSchema != null) "inputSchema": inputSchema,
        if (outputSchema != null) "outputSchema": outputSchema,
      };

  @override
  bool operator ==(Object other) =>
      identical(this, other) || other is ToolDefinition && name == other.name && description == other.description;

  @override
  int get hashCode => Object.hash(name, description);

  @override
  String toString() => "ToolDefinition(name: $name, description: $description)";
}
